### DISCUSSION TITLE
nginx config validation

### DISCUSSION THREAD

**Discussion by cpontvieux-systra** [nginx-proxy/nginx-proxy#1898] [Ideas]
Under some (unfortunate) circumstances, it could be possible to generate a wrong nginx configuration.

The problem is, for now, that the configuration is generated on the fly directly in `/etc/nginx` directories.

How about generating configuration in alternate directories, and then check the configuration using `nginx -t -c /etc/main-nginx.conf` prior to modifying `/etc/nginx` config files.

This would ensure that the whole config is ok before nginx got it, and that no erroneous configs are written to disk.

This is just a potential risk but if a bad config is written, nginx will not load it (because it’s wrong). But the next time it will restart, the on-disk bad config will be read and nginx will crash.

This is tricky, because from a first quick read, this may impact `functions.sh`, `start.sh` and potentially others…

**Comment by marc-hg** [comment]
Not directly related to your post but do you know how to change the nginx default.conf? I'm trying to add client_max_body_size to be able to upload files and it just gets overwritten when I docker-compose up

**Reply by buchdag** [reply]
@marc-hg if you have a specific question that's not directly related to the issue's or discussions's topic, I'd prefer that you ask it separately on the [Discussions section](https://github.com/nginx-proxy/nginx-proxy/discussions). :slightly_smiling_face:

**Comment by buchdag** [comment]
@cpontvieux-systra that's a very interesting idea but it would require some work. Right now the docker-gen process render the nginx conf from the template then signal the nginx to reload its config. We could instead generate the nginx conf to a temp file, then wrap the following in a script: 

- verify the temporary config with `nginx -t`
- if it's not okay log a warning then do nothing
- if it's okay write the temp config to `/etc/nginx` then issue `nginx -s reload`

And have the docker-gen  process call this script instead of directly notifying nginx.

However I don't really see how that would work for the two containers version (separate nginx and docker-gen containers).
